### PR TITLE
Bottom sheet component

### PR DESCRIPTION
**Pull request by rasom**
## how to test
1. log in 
2. press status logo on home screen (chats list)
3. close it or something
4. press status logo again if you want to see the sheet again 

## samsung a500h:
bots2

### CODE REVIEW

File: src/status_im/ui/components/bottom_sheet/view.cljs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+(def initial-animation-duration 300)
+(def release-animation-duration 100)
+(def cancellation-animation-duration 150)
+(def swipe-opacity-range 100)
+(def cancellation-height 250)
+
+(defn animate-sign-panel
+  [opacity-value bottom-value]
+  (animation/start
```
Comment (mandrigin) [inline src/status_im/ui/components/bottom_sheet/view.cljs:19]: would this one use a native driver? otherwise it might be a bit choppy on slower devices...

Comment (rasom) [reply src/status_im/ui/components/bottom_sheet/view.cljs:19]: fixed

**Comment by rasom** [comment]
@andmironov you may want to check it

File: env/dev/env/config.cljs [hunk 2]
```diff
@@ -1,6 +1,6 @@
 (ns env.config)
 
 (def figwheel-urls {:android "ws://localhost:3449/figwheel-ws",
- :ios "ws://192.168.0.9:3449/figwheel-ws",
+ :ios "ws://10.0.1.4:3449/figwheel-ws",
```
Comment (mandrigin) [inline env/dev/env/config.cljs]: I think this one is committed by mistake.

I wonder how can we avoid changing this file all the time... maybe we can read from an env variable?

**Review by mandrigin — CHANGES_REQUESTED** [review]
looks like a leftover change from debugging.
also, we probably shoudn't merge it with this test touch handler.

File: env/dev/env/config.cljs (hunk 2 above)
Comment (cammellos) [reply env/dev/env/config.cljs]: I think we can just add this to gitignore?

**Comment by flexsurfer** [comment]
@rasom could you please provide some description about the purpose of this component? thanks

File: src/status_im/ui/components/bottom_sheet/view.cljs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+   :align-self      :stretch
+   :justify-content :center
+   :align-items     :center})
+
+(def handle-style
+  {:width            31
+   :height           4
+   :background-color colors/gray-transparent-40})
+
+(def bottom-view-style
```
Comment (flexsurfer) [inline src/status_im/ui/components/bottom_sheet/view.cljs]: usually, we move styles in separate ns?

File: src/status_im/ui/components/colors.cljs [hunk 4]
```diff
@@ -22,10 +22,12 @@
 ;; BLACK
 (def black "#000000") ;; Used as the default text color
 (def black-transparent (alpha black 0.1)) ;; Used as background color for rounded button on dark background and as background color for containers like "Backup seed phrase"
+(def black-transparent-40 (alpha black 0.4))
```
Comment (flexsurfer) [inline src/status_im/ui/components/colors.cljs:25]: @denis-sharypin new colour?

File: env/dev/env/config.cljs (hunk 2 above)
Comment (flexsurfer) [reply env/dev/env/config.cljs]: let's finally do it :)

File: src/status_im/ui/components/bottom_sheet/view.cljs (hunk 3 above)
Comment (rasom) [reply src/status_im/ui/components/bottom_sheet/view.cljs]: fixed

File: env/dev/env/config.cljs (hunk 2 above)
Comment (rasom) [reply env/dev/env/config.cljs]: fixed

File: src/status_im/ui/components/bottom_sheet/view.cljs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                        :useNativeDriver true})
+     (animation/timing bottom
+                       {:toValue         new-bottom-value
+                        :duration        duration
+                        :useNativeDriver true})])
+   (when (fn? callback) callback)))
+
+(defn animate-sign-panel
+  [opacity-value bottom-value]
+  (animate {:bottom   bottom-value, :new-bottom-value 0
```
Comment (dmitryn) [inline src/status_im/ui/components/bottom_sheet/view.cljs]: minor styling comment: you don't need comma unless you use it to visually separate key value pairs

File: src/status_im/ui/components/bottom_sheet/view.cljs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+            :opacity  opacity-value, :new-opacity-value 1
+            :duration initial-animation-duration}))
+
+(defn- on-move
+  [{:keys [height bottom-value opacity-value]}]
+  (fn [_ state]
+    (let [dy (.-dy state)]
+      (cond (pos? dy)
+            (let [opacity (max 0.05 (- 1 (/ dy (- height swipe-opacity-range))))]
```
Comment (dmitryn) [inline src/status_im/ui/components/bottom_sheet/view.cljs]: 0.05 looks like magic number

File: src/status_im/ui/components/bottom_sheet/view.cljs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+              (animation/set-value bottom-value dy)
+              (animation/set-value opacity-value opacity))
+            (neg? dy)
+            (animation/set-value bottom-value (/ dy 2))))))
+
+(defn cancelled? [height dy vy]
+  (or
+   (< 0.1 vy)
```
Comment (dmitryn) [inline src/status_im/ui/components/bottom_sheet/view.cljs]: one more magic number. Also `dy` `vy` are very short names, hard to understand what do they mean without context

**Review by dmitryn — APPROVED** [review]

File: src/status_im/ui/components/bottom_sheet/view.cljs (hunk 5 above)
Comment (rasom) [reply src/status_im/ui/components/bottom_sheet/view.cljs]: matter of taste, if that's an issue i can make one kv pair per line 

https://github.com/bbatsov/clojure-style-guide#opt-commas-in-map-literals

File: src/status_im/ui/components/bottom_sheet/view.cljs (hunk 7 above)
Comment (rasom) [reply src/status_im/ui/components/bottom_sheet/view.cljs]: that's names from PanResponder API so idk, not sure we need to invent it again https://facebook.github.io/react-native/docs/panresponder

Comment (dmitryn) [reply src/status_im/ui/components/bottom_sheet/view.cljs]: ok, makes sense not to change them

File: src/status_im/ui/components/bottom_sheet/view.cljs (hunk 5 above)
Comment (dmitryn) [reply src/status_im/ui/components/bottom_sheet/view.cljs]: not an issue, just noticed this as we usually don't use commas

File: src/status_im/ui/components/bottom_sheet/styles.cljs [hunk 8]
```diff
@@ -0,0 +1,54 @@
+(ns status-im.ui.components.bottom-sheet.styles
+  (:require [status-im.ui.components.colors :as colors]))
+
+(def border-radius 16)
+(def bottom-padding 34)
```
Comment (errorists) [inline src/status_im/ui/components/bottom_sheet/styles.cljs]: hey, the bottom padding in the component is specific to iPhone X and later which require a 34pt high safe area. For other devices it should be a much smaller number like 8.

File: src/status_im/ui/components/colors.cljs (hunk 4 above)
Comment (errorists) [reply src/status_im/ui/components/colors.cljs:25]: @flexsurfer :) it's a good point but honestly, we don't have a transparent colour in the current palette that would fit here, 40% grey is too light and 10% black even worse. can it be used locally?

File: src/status_im/ui/components/bottom_sheet/styles.cljs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+(def content-header
+  {:height          border-radius
+   :align-self      :stretch
+   :justify-content :center
+   :align-items     :center})
+
+(def handle
+  {:width            31
+   :height           4
+   :background-color colors/gray-transparent-40})
```
Comment (errorists) [inline src/status_im/ui/components/bottom_sheet/styles.cljs]: the handle should also include a 2px border radius, thx!

**Comment by flexsurfer** [comment]
could someone please provide some description about the purpose of this component? thanks

File: src/status_im/ui/components/bottom_sheet/styles.cljs (hunk 8 above)
Comment (rasom) [reply src/status_im/ui/components/bottom_sheet/styles.cljs]: ok, I'll change this. Is it specified anywhere in figma?

**Comment by errorists** [comment]
> could someone please provide some description about the purpose of this component? thanks

Sure, I can. From a design POV, the bottom sheet unifies and replaces the native [Action Sheet](https://developer.apple.com/design/human-interface-guidelines/ios/views/action-sheets/) view on iOS and [Bottom Sheet](https://material.io/design/components/sheets-bottom.html#anatomy) on Android. It's used for supplementary menus where after an action is made we're presenting users with a bunch of choices, a good example would be tapping the '•••' in Chats or the 'plus' button in Home.

**Comment by rasom** [comment]
@flexsurfer in my case it is needed specifically for https://github.com/status-im/status-react/issues/7288

File: src/status_im/ui/components/bottom_sheet/styles.cljs (hunk 8 above)
Comment (errorists) [reply src/status_im/ui/components/bottom_sheet/styles.cljs]: thanks! it is, [link](https://www.figma.com/file/cb4p8AxLtTF3q1L6JYDnKN15/Index?node-id=858%3A81)

Comment (rasom) [reply src/status_im/ui/components/bottom_sheet/styles.cljs]: ok, thanks. I didn't know that there is a separate page for the bottom sheet

File: src/status_im/ui/components/bottom_sheet/view.cljs (hunk 6 above)
Comment (rasom) [reply src/status_im/ui/components/bottom_sheet/view.cljs]: fixed

File: src/status_im/ui/components/bottom_sheet/view.cljs (hunk 5 above)
Comment (rasom) [reply src/status_im/ui/components/bottom_sheet/view.cljs]: fixed

File: src/status_im/ui/components/bottom_sheet/view.cljs (hunk 7 above)
Comment (rasom) [reply src/status_im/ui/components/bottom_sheet/view.cljs]: magic number fixed

File: src/status_im/ui/components/bottom_sheet/styles.cljs (hunk 8 above)
Comment (rasom) [reply src/status_im/ui/components/bottom_sheet/styles.cljs]: fixed

File: src/status_im/ui/components/bottom_sheet/styles.cljs (hunk 9 above)
Comment (rasom) [reply src/status_im/ui/components/bottom_sheet/styles.cljs]: fixed

**Comment by rasom** [comment]
@flexsurfer, @errorists, @mandrigin, @dmitryn requested changes are done

**Comment by andmironov** [comment]
The component works awesome!

Notes: 
1. the component may or may not be closable hence may or may not have that handle at the top and be swipable.

screen shot 2019-01-29 at 12 37 20


2. the component should handle different heights well

screen shot 2019-01-29 at 12 36 35

**Comment by rasom** [comment]
@andmironov i think "not closable" part will be added when needed for another use case, don't want to spend extra time on this now

**Review by flexsurfer — APPROVED** [review]

**Review by mandrigin — APPROVED** [review]

**Comment by Serhy** [comment]
@rasom Looks good to me on iPhoneX, Android 6.0.1, Android 7, Android 8. Smooth appearing/disappearing upon sweep or tap on 'Close' button in the test sheet on a weak device (Note 4). Flipped chat items on Home screen (when a message is received) doesn't have impact on opened bottom sheet. 
Ready for the merge, if I'm correct with two items below:  
1) There is white gap at the very bottom of component, especially visible on iPhoneX: 
screenshot 2019-01-30 at 13 10 06

However, it's how test bottom sheet created, i.e. when later we introduce for example same bottom sheet upon tap Status logo not for testing purposes and with "real" useful data, - it'll be a part of styling particular sheet, correct?

2) Conflicts appeared, but the conflicting code is the `[do not merge] test bottom sheet` part of the commit of this PR, which we won't merge, so it's OK.

**Comment by rasom** [comment]
@Serhy see  https://github.com/status-im/status-react/pull/7343#discussion_r251717061 regarding 1)
2) is fixed

**Comment by Serhy** [comment]
@rasom Android builds failed. Before we re-run them, could you please rebase on develop? So we have #7381 changes included here too.

**Comment by rasom** [comment]
@Serhy done

**Pull request merged**